From b30bdcdc28a94bf8d1e7075c008bfd9fd56d2896 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Fri, 24 May 2024 11:01:10 +0800
Subject: [PATCH 092/149] mtk: hostapd: fix using wrong link id in
 nl80211_set_channel during set beacon

param.freq->link_id is used in nl80211_set_channel but it is not set in __ieee802_11_set_beacon.
nl80211_set_channel will return EBUSY after channel switch.
This error occurs repeatedly especially when the bandwidth is changed.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 src/ap/beacon.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 3254c1187..a97d46f52 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2832,6 +2832,12 @@ static int __ieee802_11_set_beacon(struct hostapd_data *hapd)
 		params.freq = &freq;
 	}
 
+	params.freq->link_id = -1;
+#ifdef CONFIG_IEEE80211BE
+	if (hapd->conf->mld_ap)
+		params.freq->link_id = hapd->mld_link_id;
+#endif /* CONFIG_IEEE80211BE */
+
 	for (i = 0; i < hapd->iface->num_hw_features; i++) {
 		mode = &hapd->iface->hw_features[i];
 
-- 
2.45.2

